pkg/updatestatus/nodeinformer.go

```diff
@@ -0,0 +1,450 @@
+	// Default to drain process so that we don't report completed.
+	mcdState := node.Annotations[mco.MachineConfigDaemonStateAnnotationKey]
+	return isUpdating && mcdState == mco.MachineConfigDaemonStateDone
+}
+
+func mcdUpdatingStateToPhase(state string) nodePhase {
+	switch state {
+	case mco.MachineConfigDaemonStateWorking:
+		return phaseStateUpdating
+	case mco.MachineConfigDaemonStateRebooting:
+		return phaseStateRebooting
+	case mco.MachineConfigDaemonStateDone:
+		return phaseStateUpdated
+	// The MCD state annotation is not set
+	case "":
+		return phaseStateUpdating
+	default:
+		// For other MCD states during an update default to updating
+		return phaseStateUpdating
+	}
+}
+
+func calculatePhase(pool *machineconfigv1.MachineConfigPool, node *corev1.Node, isUpdating, isUpdated bool) nodePhase {
+	var phase nodePhase
+	switch {
+	case isUpdating && isNodeDraining(node, isUpdating):
+		phase = phaseStateDraining
+	case isUpdating:
+		phase = mcdUpdatingStateToPhase(node.Annotations[mco.MachineConfigDaemonStateAnnotationKey])
+	case isUpdated:
+		phase = phaseStateUpdated
+	case pool.Spec.Paused:
+		phase = phaseStatePaused
+	default:
+		phase = phaseStatePending
+	}
+	return phase
+}
+
+func toPointer(d time.Duration) *metav1.Duration {
+	v := metav1.Duration{Duration: d}
+	return &v
+}
+
+func assessNode(node *corev1.Node, mcp *machineconfigv1.MachineConfigPool, machineConfigs map[string]*machineconfigv1.MachineConfig, cv *configv1.ClusterVersion, now metav1.Time) *NodeStatusInsight {
+	if node == nil || mcp == nil || cv == nil {
+		return nil
+	}
+
+	desiredConfig, ok := node.Annotations[mco.DesiredMachineConfigAnnotationKey]
+	currentVersion, foundCurrent := getOpenShiftVersionOfMachineConfig(machineConfigs, node.Annotations[mco.CurrentMachineConfigAnnotationKey])
+	desiredVersion, foundDesired := getOpenShiftVersionOfMachineConfig(machineConfigs, desiredConfig)
+
+	lns := mco.NewLayeredNodeState(node)
+	isUnavailable := lns.IsUnavailable(mcp)
+
+	isDegraded := isNodeDegraded(node)
+	isUpdated := foundCurrent && isLatestUpdateHistoryVersionEqualTo(cv.Status.History, currentVersion) &&
+		// The following condition is to handle the multi-arch migration because the version number stays the same there
+		(!ok || node.Annotations[mco.CurrentMachineConfigAnnotationKey] == desiredConfig)
+
+	// foundCurrent makes sure we don't blip phase "updating" for nodes that we are not sure
+	// of their actual phase, even though the conservative assumption is that the node is
+	// at least updating or is updated.
+	isUpdating := !isUpdated && foundCurrent && foundDesired && isLatestUpdateHistoryVersionEqualTo(cv.Status.History, desiredVersion)
+
+	condition := metav1.Condition{
+		Type:               string(NodeStatusInsightUpdating),
+		Status:             metav1.ConditionUnknown,
+		Reason:             string(NodeCannotDetermine),
+		LastTransitionTime: now,
+	}
+
+	phase := calculatePhase(mcp, node, isUpdating, isUpdated)
```

Comment:
Hmm... I am not sure whether I like the indirection where we first compute a "phase" which is not a concept present in the API, and then we build a kube condition from  :thinking: 

Would it make sense to compute the condition content right away `calculateX` method, without the intermediate `phase`?

Reply:
Done